src/mpid/ch4/shm/ipc/src/ipcg_p2p.h

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef IPCG_P2P_H_INCLUDED
+#define IPCG_P2P_H_INCLUDED
+
+#include "ch4_impl.h"
+#include "shm_control.h"
+#include "ipc_pre.h"
+#include "ipc_impl.h"
+#include "ipc_mem.h"
+
+/* Generic IPC protocols for P2P. */
+
+/* Generic sender-initialized LMT routine with contig send buffer.
+ *
+ * If the send buffer is noncontiguous the submodule can first pack the
+ * data into a temporary buffer and use the temporary buffer as the send
+ * buffer with this call. The sender gets the memory handle of the specified
+ * buffer, and sends to the receiver. The receiver will then open the remote
+ * memory handle and perform direct data transfer.
+ */
+MPL_STATIC_INLINE_PREFIX int MPIDI_IPCG_send_contig_lmt(const void *buf, MPI_Aint count,
+                                                        MPI_Datatype datatype, int rank, int tag,
+                                                        MPIR_Comm * comm, int context_offset,
+                                                        MPIDI_av_entry_t * addr,
+                                                        MPIDI_SHM_IPC_type_t ipc_type,
+                                                        MPIR_Request ** request)
+{
```

Comment:
Since we only deal with contig types here, we don't need `count` and `datatype` arguments to be passed. The submodule calling the generic function would just pass `data_sz`.

Reply:
Makes sense.

Reply:
Hmm thinking better, maybe not :) This function fills in generic request information that is shared by all submodules which includes datatype and count. Anyway, nevermind my comment. I think this is fine for now.